### ISSUE TITLE
Finding unknown block types

### ISSUE THREAD

**Issue by Cylindric** [mapcrafter/mapcrafter#243]
Hi. I'm running a server with the DireWolf20 pack on it, so there are a whole bunch of missing textures.
Is there any way of getting mapcrafter to output the name or ID of the textures it is skipping, so I can find and add them?

**Comment by Cylindric** [comment]
Actually, research indicates that might all be tricky with mods, so would it be possible instead to just substitute a different block for "unknown"? It would be good to put in a default stone block, for example, rather than leaving weird floating gaps and roofs etc.

**Comment by m0r13** [comment]
Oh hey! There is actually an option to highlight unknown blocks. Just put a `render_unknown_blocks = true` into your map section.

**Comment by Cylindric** [comment]
Yeah, that makes them bright red though :) It would be good to get them
something like 'stone', so the map looks more natural.
I have actually hacked the source and just changed the "unknown block"
renderer to use the STONE block instead. It's not clever, but it looks
better than holes (and therefore floating structures) or bright red blocks.

I think a related problem is the cave renderer. It seems to render all
unknown blocks as air, and hence tunnels, when in fact they are far more
likely to be solid.

**Comment by C0rish** [comment]
I'm stucking on the same problem... but with a windows client.
There is no chance for me to edit the code and change the "red" color to something else.

The DireWolf20 Server looks a little bit crazy with missing blocks and a lot more crazy with a whole bunch of red dots...

Is there a "easy" way (remember: windows) to change the color or export a list of unknown ids? And then... reimport ids with textures?

**Comment by Cylindric** [comment]
I edited the code to output a message for every missed block ID. Unfortunately the block IDs didn't match anything I found in-game, so the only thing I could do was render a tiny portion of the map with the offending blocks in it, and then edit those into the code with the closest vanilla block and re-build again.

I forked the project here: [Cylindric/mapcrafter](https://github.com/Cylindric/mapcrafter/)

And here is the commit that adds the block ID output: [59adbf](https://github.com/Cylindric/mapcrafter/commit/59adbf61326cd9700af19d031116cc3d5dee9a1e)

Here's a sample of a new block: [3738ff](https://github.com/Cylindric/mapcrafter/commit/3738ff3c607811bdcb27cc02c6de1cbea936888d#diff-2f400f005bdd1ecdcad1746aec8329f1R1884)

It's a fairly onerous process as it stands :) I have been:

1. Find an interesting part of the map with missing blocks (someone's base)
2. Run my build of mapcrafter and see which blocks show up a lot
3. Add a new entry in the code for that block Id, set to something obvious like WOOL_COLORED_CYAN or SPONGE
4. Build and run mapcrafter again, look at the map to find the wool blocks, and identify them in-game
5. Amend the new entry from #3 with the closest matching vanilla MC value
6. Repeat until it looks "good enough".

I have a little one-liner I use to build then run mapcrafter, that counts all the blocks and outputs them in order of how many times they are present - I just target the ones with loads:

`make && ./src/mapcrafter --config test/mapcrafter.conf --render-force-all | grep block\ for | cut -d ' ' -f 9,10 | sort | uniq -c | sort -n
`

I still need to work out how to add totally new textures to it :)

**Comment by m0r13** [comment]
Hehe nice hack. Can't you also use F3 in-game to find out block ID/data?

I wrote a little bit about custom blocks in another issue: https://github.com/mapcrafter/mapcrafter/issues/217#issuecomment-207828003

You might want to be generously with the block data first (configure a data mask to ignore all block data for your new blocks) so all new blocks with a specific id are matched. Then you can continue on figuring out the block data

**Comment by Cylindric** [comment]
I need to try F3 again, but from my first test, the numbers shown in-game do not in any way correspond to the values in mapcrafter.  I'll have to try again, and post some examples.

**Comment by C0rish** [comment]
A config file for additional blocks would be nice... Together with an output error log telling something like "block XYZ missing, please specify in config file".

In this case you would be able to set block ID XYZ to resource name XYZ :)
1234 = mod/PNG/graphics.png
1357 = mod/PNG/anothergrapics.png

Just a little idea to make it user friendly.

**Comment by C0rish** [comment]
Just so clear things up:

Default rendering like it is - don't show unknown blocks - is okay, but then followed with an "please specify ID" error in a log file.

After that every additional, specified block will be also rendered, when you put it in this config file.

Couldn't be that much change in the code or?

**Comment by m0r13** [comment]
@C0rish: Of course it would be nice to have some kind of block image configuration - that would also make support for mod block easier.

Unfortunately the management of block images is more complicated than you might think. The block images are dynamically generated depending on the textures / texture size / render view / extra block data. It might be possible to create just a partial block configuration (for the simple cubical blocks) and have other block images still generated by custom code, but I really don't have the time to work on that anymore as I'm busy with other projects. Sorry.

**Comment by sethborg** [comment]
Cylindric, can your fork drop a pin on the spot on the map where a specific block type is located? I'd love to see that feature. Config might be: pin_block_id: 56,58 // diamonds, crafting tables

**Comment by Cylindric** [comment]
No, it changes the texture used to render the unknown block. I don't create any files or pins or anything. That doesn't really have anything to do with this.

**Comment by Fenrirthviti** [comment]
Greetings! I realize this is quite an old issue, but I've run into the same problems mentioned here. While adding each individual block is something I don't have much interest in, could someone point me in the right direction to change how the "unknown" block is rendered? I attempted to change the color of the blocks, which I traced back to (I assumed?) [/src/mapcraftercore/mc/chunk.cpp#L327-L330](https://github.com/mapcrafter/mapcrafter/blob/master/src/mapcraftercore/mc/chunk.cpp#L327-L330).

I tried changing the value in the same file at [#L280-L286](https://github.com/mapcrafter/mapcrafter/blob/master/src/mapcraftercore/mc/chunk.cpp#L282) but seems that was not correct. I'd like to just render ALL unknown blocks to a custom texture file, or at least some way of changing the color the unknown blocks are displayed as, since red is a bit glaring, and something like a light grey or off-white would be just fine for my use case.

**Comment by mostlyfabulous** [comment]
@Cylindric This is late but I found out that the Block ID & state you probably want (especially for mod blocks) can be found by using `/info` in-game and this will bind the current thing you're holding in Minecraft to display info about all solid blocks you right click.

The F3 debugger mode has a Block ID that corresponds to some other Block ID in NBT files.
